File: src/Heap/lastStone.js
/*
Last Stone Weight
You are given an array of integers stones where stones[i] represents the weight of the ith stone.

We want to run a simulation on the stones as follows:

At each step we choose the two heaviest stones, with weight x and y and smash them togethers
If x == y, both stones are destroyed
If x < y, the stone of weight x is destroyed, and the stone of weight y has new weight y - x.
Continue the simulation until there is no more than one stone remaining.

Return the weight of the last remaining stone or return 0 if none remain.

Example 1:

Input: stones = [2,3,6,2,4]

Output: 1
*/

class Solution {
  /**
   * @param {number[]} stones
   * @return {number}
   */
  lastStoneWeight(stones) {
    while (stones.length > 1) {
      // Sort in descending order to get the two heaviest stones
      stones.sort((a, b) => b - a);
      const y = stones.shift(); // largest
      const x = stones.shift(); // second largest

      if (y !== x) {
        stones.push(y - x);
      }
      // else both are destroyed, do nothing
    }

    return stones[0] || 0; // return last remaining stone or 0
  }
}

File: src/Heap/lastStone.py
import heapq
from typing import List

"""
You are given an array of integers stones where stones[i] is the weight of the ith stone.
We are playing a game with the stones. On each turn, we choose the heaviest two stones and smash them together. Suppose the heaviest two stones have weights x and y with x <= y. The result of this smash is:
If x == y, both stones are destroyed, and
If x != y, the stone of weight x is destroyed, and the stone of weight y has new weight y - x.
At the end of the game, there is at most one stone left.
Return the weight of the last remaining stone. If there are no stones left, return 0.
"""


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        max_heap = [-n for n in stones]
        heapq.heapify(max_heap)
        while len(max_heap) > 1:
            # while the heap has more than 1 element
            max_val1 = -heapq.heappop(max_heap)
            max_val2 = -heapq.heappop(max_heap)
            diff = max_val1 - max_val2
            # if the diff of top two is more than zero, then push it to the heap
            if diff > 0:
                heapq.heappush(max_heap, -diff)
        # after there is 1 or 0 element in the heap
        return -heapq.heappop(max_heap) if max_heap else 0
